docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\externTyp \name$ | $\qquad \text{externs}$ |
+|   | $\vert$ | $\funcType \kind \typParams \params \typ$ | $\qquad \text{functions}$ |
+|   | $\vert$ | $\actionTyp {\overline {\prm 1}} {\overline {\prm 2}}$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\constructorTyp \typParams \prms \wildcardParams \typ$ | $\qquad \text{constructors}$ |
+|   | $\vert$ | $\tableTyp \name$ | $\qquad \text{tables}$ |
+{ .booktable }
+
+
+**Directions:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\direction$ | $::=$ | $\inDir$ | $\qquad \text{copy-in}$ |
+|   | $\vert$ | $\out$ | $\qquad \text{copy-out}$ |
+|   | $\vert$ | $\inout$ | $\qquad \text{copy-in-out}$ |
+|   | $\vert$ | $\less$ | $\qquad \text{direction-less}$ |
+
+**Parameters:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\prmDef$ | $::=$ | $\dir \, \var : \baseTyp$ | $\qquad \text{parameters}$ |
+|   | $\vert$ | $\dir \,\var : (\baseTyp, \exp)$ | $\qquad \text{parameters with value}$ |
+
+**Unary operations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\unaryOp$ | $::=$ | $!$ | $\qquad \text{logical negation}$ |
+|   | $\vert$ | $\bitComplement$ | $\qquad \text{bitwise complement}$ |
+|   | $\vert$ | $-$ | $\qquad \text{unary minus}$ |
+
+**Binary operations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\binOp$ | $::=$ | $\&\&$ | $\qquad \text{logical and}$ |
+|   | $\vert$ | $\vert\vert$ | $\qquad \text{logical or}$ |
+|   | $\vert$ | $+$ | $\qquad \text{numerical addition}$ |
+|   | $\vert$ | $-$ | $\qquad \text{numerical subtraction}$ |
+|   | $\vert$ | $*$ | $\qquad \text{numerical multiplication}$ |
+|   | $\vert$ | $\div$ | $\qquad \text{numerical division}$ |
+|   | $\vert$ | $\mod$ | $\qquad \text{numerical modulo}$ |
+|   | $\vert$ | $==$ | $\qquad \text{equality check}$ |
+|   | $\vert$ | $!=$ | $\qquad \text{inequality check}$ |
+|   | $\vert$ | $\plusSat$ | $\qquad \text{saturating addition}$ |
+|   | $\vert$ | $\subSat$ | $\qquad \text{saturation subtraction}$ |
+|   | $\vert$ | $\bitAnd$ | $\qquad \text{bitwise and}$ |
+|   | $\vert$ | $\bitOr$ | $\qquad \text{bitwise or}$ |
+|   | $\vert$ | $\bitXor$ | $\qquad \text{bitwise xor}$ |
+|   | $\vert$ | $\bitComplement$ | $\qquad \text{bitwise complement}$ |
+|   | $\vert$ | $\concat$ | $\qquad \text{bitwise concatenation}$ |
+|   | $\vert$ | $\shiftL$ | $\qquad \text{shift left}$ |
+|   | $\vert$ | $\shiftR$ | $\qquad \text{shift right}$ |
+|   | $\vert$ | $<$ | $\qquad \text{less than}$ |
+|   | $\vert$ | $\leq$ | $\qquad \text{less than or equal}$ |
+|   | $\vert$ | $>$ | $\qquad \text{greater than}$ |
+|   | $\vert$ | $\geq$ | $\qquad \text{greater than or equal}$ |
+
+**Arguments:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\arg$ | $::=$ | $\argexp$ | $\qquad \text{expression}$ |
+|   | $\vert$ | $\argkv$ | $\qquad \text{key value}$ |
+|   | $\vert$ | $\missingarg$ | $\qquad \text{missing}$ |
```

Comment:
Here $\missingarg$ renders as "_" whereas previous "don't care" types and directionless parameters were written out in the abstract syntax. Let's be consistent?

Reply:
applied.

Reply:
- [] discuss in meeting: arguments in P4 can also be `x = dontcare` but Petr4 doesn't allow this. Include it in abstract syntax?

Reply:
@pataei double check this and add it to issues, if applicable.